File: python3/best_time_stock1.py
"""
Space   : O(1)
Time    : O(n)
"""
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        start, dp = 10**10, 0
        
        for i in prices:
            print(start)
            start = min(start, i)
            dp = max(dp, i-start)

        return dp
        

File: golang/best_time_stock_1.go
package main

// Space   : O(1)
// Time    : O(n)

func maxProfit(prices []int) int {
	start := 10000000
	dp := 0

	for _, v := range prices {
		if v < start {
			start = v
		}
		if dp < v-start {
			dp = v - start
		}
	}

	return dp
}
